ksql-common/src/main/java/io/confluent/ksql/util/KsqlConfig.java

```diff
@@ -189,115 +240,208 @@
+        .forEach(
+            resolvedConfig -> props.put(resolvedConfig.key, resolvedConfig.value));
+    return Collections.unmodifiableMap(props);
   }
 
-  public ImmutableMap<String, Object> getAllProps() {
-    final ImmutableMap.Builder<String, Object> builder = ImmutableMap.builder();
-    builder.putAll(values().entrySet().stream()
-        .filter(e -> e.getValue() != null)
-        .collect(
-            Collectors.toMap(
-                Map.Entry::getKey,
-                Map.Entry::getValue))
+  public Map<String, String> getKsqlConfigPropsCleaned() {
+    final Map<String, String> props = new HashMap<>();
+    configDef(true).names().stream().forEach(
+        key -> props.put(key, ConfigDef.convertToString(values().get(key), typeOf(key)))
     );
-    builder.putAll(ksqlStreamConfigProps.entrySet().stream()
-        .filter(e -> e.getValue() != null)
-        .collect(
+    return Collections.unmodifiableMap(props);
+  }
+
+  public Map<String, String> getKsqlStreamConfigPropsCleaned() {
+    final Map<String, String> props = new HashMap<>();
+    ksqlStreamConfigProps.values().forEach(
+        resolvedConfig -> props.put(
+            resolvedConfig.key,
+            ConfigDef.convertToString(resolvedConfig.value, resolvedConfig.type))
+    );
+    return Collections.unmodifiableMap(props);
+  }
+
+  public Map<String, String> getAllConfigPropsCleaned() {
+    final Map<String, String> allPropsCleaned = new HashMap<>();
+    allPropsCleaned.putAll(getKsqlConfigPropsCleaned());
+    allPropsCleaned.putAll(
+        getKsqlStreamConfigPropsCleaned().entrySet().stream().collect(
             Collectors.toMap(
-                e -> KSQL_STREAMS_PREFIX + e.getKey(),
-                Map.Entry::getValue)));
-    return builder.build();
+                e -> KSQL_STREAMS_PREFIX + e.getKey(), Map.Entry::getValue
+            )
+        )
+    );
+    return Collections.unmodifiableMap(allPropsCleaned);
   }
 
   public KsqlConfig clone() {
-    return new KsqlConfig(values(), ksqlStreamConfigProps);
+    return new KsqlConfig(true, values(), ksqlStreamConfigProps);
   }
 
   public KsqlConfig cloneWithPropertyOverwrite(final Map<String, Object> props) {
     final Map<String, Object> cloneProps = new HashMap<>(values());
     cloneProps.putAll(props);
-    final Map<String, Object> streamsConfigOverlay = new HashMap<>();
-    applyStreamsConfig(ksqlStreamConfigProps, streamsConfigOverlay);
-    applyStreamsConfig(props, streamsConfigOverlay);
-    return new KsqlConfig(cloneProps, ImmutableMap.copyOf(streamsConfigOverlay));
+    final Map<String, ResolvedConfig> streamConfigProps = new HashMap<>();
+    applyStreamsConfig(getKsqlStreamConfigProps(), streamConfigProps);
+    applyStreamsConfig(props, streamConfigProps);
+    return new KsqlConfig(true, cloneProps, ImmutableMap.copyOf(streamConfigProps));
+  }
+
+  public KsqlConfig mergeWithOriginalConfig(final Map<String, String> props) {
```

Comment:
The purpose of this method is to preserve the original values for configs which break compatibility. Maybe a better name would be "overrideBreakingConfigsWithOriginalValues".